cmd/server/bookmark/repository.go

```diff
@@ -0,0 +1,63 @@
+package bookmark
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+)
+
+type Repository struct {
+	DB *sql.DB
+}
+
+func NewRepository(db *sql.DB) *Repository {
+	return &Repository{DB: db}
+}
+
+// Transactionally executes a function within a database transaction. It commits the transaction
+// if the function succeeds, otherwise it rolls back. If rollback fails, both errors are returned.
+func (r *Repository) Transactionally(ctx context.Context, f func(tx *sql.Tx) error) error {
+	tx, err := r.DB.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	err = f(tx)
+	if err != nil {
+		if rbErr := tx.Rollback(); rbErr != nil {
+			return fmt.Errorf("transaction rollback error: %w, original error: %w", rbErr, err)
+		}
+		return err
+	}
+
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateBookmark creates a new bookmark.
+func (r *Repository) CreateBookmark(ctx context.Context, tx *sql.Tx, bookmark *Bookmark) (*Bookmark, error) {
```

Comment:
We can consider to pass the bookmark by value, because this makes it immutable and prevents some unintended side effects.

Reply:
Hmm, we could do that. However I thought it was generally more advisable to pass along pointers to asure you do not create copies. Also we modify the bookmark in the repository function so we would need to pass in a copy and return a modified copy. What are you worries about unintented side effects?

Reply:
Creating copies is 99% of the time fine in go. The Go’s compiler is pretty good at optimizing small struct copies, so the performance impact is often minimal in such cases. By returning a modified Bookmark and treating the input as immutable, you make the contract of the function more explicit. The caller knows that the input is never modified directly, and if they need the modified version, they will work with the returned value. 

As for unintended side effects, if you pass a pointer, you risk allowing changes to Bookmark to happen in the calling context without that being clear. 

I'm also fine with an pointer, but then i would consider:
```go
func (r *Repository) CreateBookmark(ctx context.Context, tx *sql.Tx, bookmark *Bookmark) error {
```